apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableServiceApiClientPartial.cs

```diff
@@ -133,4 +187,383 @@ public partial class BigtableServiceApiClientImpl
                 "Non-idempotent MutateRows requests are not allowed. Specify a version with all SetCell mutations.");
     }
+
+    // TODO: Move changes into Google.Api.Gax.Grpc.ChannelPool if approved.
+    /// <summary>
+    /// A pool of channels for the same service, but with potentially different endpoints. Each endpoint
+    /// has a single channel. All channels created by this pool use default application credentials.
+    /// This class is thread-safe.
+    /// </summary>
+    public sealed class ChannelPool
+    {
+        #region NewStuff
+
+        private readonly Dictionary<GcpCallInvokerKey, GcpCallInvoker> _gcpCallInvokers = new Dictionary<GcpCallInvokerKey, GcpCallInvoker>();
+
+        private class ChannelOptionComparer : IEqualityComparer<ChannelOption>
+        {
+            public static readonly ChannelOptionComparer Instance = new ChannelOptionComparer();
+
+            private ChannelOptionComparer() { }
+
+            public bool Equals(ChannelOption x, ChannelOption y)
+            {
+                if (x == null && y == null)
+                {
+                    return true;
+                }
+
+                if (x == null || y == null)
+                {
+                    return false;
+                }
+
+                if (x.Type != y.Type || x.Name != y.Name)
+                {
+                    return false;
+                }
+
+                switch (x.Type)
+                {
+                    case ChannelOption.OptionType.Integer:
+                        return x.IntValue == y.IntValue;
+
+                    case ChannelOption.OptionType.String:
+                        return y.StringValue == y.StringValue;
+
+                    default:
+                        throw new ArgumentException("Unexpected channel option type: " + x.Type);
+                }
+            }
+
+            public int GetHashCode(ChannelOption obj)
+            {
+                return
+                    obj == null ? 0 :
+                    obj.Type == ChannelOption.OptionType.Integer ?
+                        EqualityHelpers.CombineHashCodes(obj.IntValue, obj.Name.GetHashCode(), (int)obj.Type) :
+                    obj.Type == ChannelOption.OptionType.String ?
+                        EqualityHelpers.CombineHashCodes(obj.StringValue.GetHashCode(), obj.Name.GetHashCode(), (int)obj.Type) :
+                    throw new ArgumentException("Unexpected channel option type: " + obj.Type);
+            }
+        }
+
+        private struct GcpCallInvokerKey : IEquatable<GcpCallInvokerKey>
+        {
+            public readonly ServiceEndpoint Endpoint;
+            public readonly List<ChannelOption> Options;
+            public readonly int _hash;
+
+            public GcpCallInvokerKey(ServiceEndpoint endpoint, List<ChannelOption> options)
+            {
+                Endpoint = endpoint;
+                Options = options;
+                _hash = EqualityHelpers.CombineHashCodes(
```

Comment:
Any reason to do this? I'd expect GetHashCode to only be called once by Dictionary anyway.

Reply:
Yeah you're probably right. Just a case of premature optimization.  Removed.